WIP: Create a more-info footer with several tabs

CrossEye opened the pull request:
This is a follow-up to #8992.  Please review the discussion there first.

This still has the original logic from that PR of a `related` field containing a manually-curated list of tiddlers.  It now formats the result as part of a new tabbed footer, which also includes three other tabs.  These tabs include: a list of items linking to the current tiddler (with a text link or their own `related` field, a list of items tagged with the current tiddler, and the `Info > Sources` tab.

The other tabs are based on pre-existing structures, but the Related tab will only grow useful as we manually add links to related tiddlers.

### Comments ###

  * Reusing the Sources tab seems to make a great deal of sense.  But it makes me wonder what other ones we might want to repurpose.  (I didn't do this with `Tagging` because the existing tiddler uses `<<lingo>>` and I didn't want to go there.
  * The look and feel of the footer probably needs work.  I wanted a border or color difference to set it apart from the surroundings.  But nothing I've found looks *quite* right.  I would definitely consider trying to make this fill up the bottom of the tiddler with no white space to its left, right, or bottom.  But I haven't tried that yet.
  * The icon for the `Sources` tab is not a great fit, but it was the best I could come up with from the existing icons.  Suggestions quite welcome!

It's bedtime, and I didn't finish my list of comments.  I'll try to add to it when my brain isn't trying to dream as I type!

But in the meanwhile, I would appreciate any feedback.

![image](https://github.com/user-attachments/assets/d31a2081-b671-4175-8787-19b17f829975)

![image](https://github.com/user-attachments/assets/7be1cd7f-8476-4b22-b55d-1106d3925015)

![image](https://github.com/user-attachments/assets/1a3327d8-d54b-45f4-8a4a-c1e8b0e12477)

![image](https://github.com/user-attachments/assets/c2d92855-eb4e-45ef-8d27-dbb442559c48)



Closes #8992

pmario commented on editions/tw5.com/tiddlers/Core Messages.tid:
```diff
@@ -1,7 +1,9 @@
 created: 20240421144407522
-modified: 20240422091247905
+modified: 20250326000209111
```
The **modified** field should not be changed, because for a massive change like this one, all toddlers will end up in the Recent tab for one day. 

So the modified field should be reset to the old dates.

CrossEye replied:
That's easily fixed.  I'll try to do it tonight, although it's possible that I won't be able to get to it until the weekend.

But I am curious how people tend to go about making large-ish changes.  I created a branch off master and simply fired up the Dev environment, making my changes inside the running wiki.  That by default resets the `modified` field of every tiddler I touch. I did run the simple ones through `<$action-setfield>`, and I can do that with `$timestamp = "no"`, but all the manual changes would still show up as modifications.  In fact, in this case, there were only a few additional tiddlers, and three (I think) additional existing tiddlers that I changed.  Those are all system tiddlers, where maybe this is not such a big concern. 

But for those doing a big change, do you operate directly on the `.tid` files for the finer control?  Or is there another way to avoid making these `modified` changes?

pmario commented:
I think that's a good start.

pmario commented:
> That by default resets the modified field of every tiddler I touch.

There is a switch "Timestamps Off" **before** you edit a new tiddler. 

![grafik](https://github.com/user-attachments/assets/65f0c947-1ffc-4bf2-9f32-622a3ca491a2)

yaisog commented:
Hi @CrossEye, very nice work getting this implemented so quickly. I like the Sources tab being in there, too. And the idea of showing the "reverse related" tiddlers together with the backlinks makes a lot of sense.

The styles and wording will probably be a topic of discussion which I'll try to stay out of. I'd prefer a somehow more "modern" look, but don't ask me how to achieve that. Especially the tabs. The whole site design is 14 years old, after all. But I'm going off on a tangent...

One thing I'd like to change is the tab icons. Tiddlywiki.com does not use icons in any of the tabs instances (that I know of). For the sake of consistency, this should remain so. The only way to make this worse is like the screenshot that Mario posted in the old PR which had _only_ icons. From looking at that I'd have no idea what they do (except maybe the one for "comments" if it is indeed for comments...)

The columns can be a space-saver, but they're still a bit wonky:
![{4933FB53-769A-4231-A6CA-F6698C29FBD4}](https://github.com/user-attachments/assets/875d1109-99e0-431c-ade9-40418e7cdcba)
(GettingStarted)
The second item breaks on a new line, which is shown to its right, making for weird spacing.
![{19D918A3-A946-45DE-816F-01082F9B4CC8}](https://github.com/user-attachments/assets/90c9ede4-fcac-42a1-bdc7-5fe62dfa1baa)
(HelloThere)
The second item breaks on a new line. This looks identical to the list of two separate items on the left and right. There should be some visual clue that it is a single item.

CrossEye commented:
@pmario:

> There is a switch "Timestamps Off" before you edit a new tiddler. That will keep the "old" modified value.

Ahh, perfect.  Thank you.

@yaisog:

> One thing I'd like to change is the tab icons. Tiddlywiki.com does not use icons in any of the tabs instances (that I know of). For the sake of consistency, this should remain so. 

That's trivial to fix, but I'll wait to see if there are other reaction first.  I added them because of something similar to 

> I'd prefer a somehow more "modern" look, but don't ask me how to achieve that.

I just found the plain tabs too boring.  This was simply a way to spice them up a bit.  So I want to see what others think, although perhaps I'll post another version without them.

> The columns can be a space-saver, but they're still a bit wonky

I'll see what I can do.  I really didn't like the bullets in there.  Maybe a hanging indent?  If worse comes to worst, I can simply remove the columns, which would also fix one other potential issue: if the `emptyMessage` text is wider than one column, it wraps and looks decidedly odd.  

> The styles and wording will probably be a topic of discussion which I'll try to stay out of.

I hope you won't.  Your feedback on such issues has been very helpful already.

I'm still trying to find a good way to contrast this footer with other tiddler body content.  The blue background was a simple way to do that, but I don't particularly like it.  And I'm still not sure what I would want.  Suggestions would be very welcome!

yaisog commented:
This is how Wikipedia does it after its redesign:
![{86BB450B-5157-4776-8AFE-17E14C0A5C18}](https://github.com/user-attachments/assets/b9ff3f15-9a83-40de-b39d-3a3762385a3f)
Maybe this can be an inspiration, though they don't have tabs.
Make the tabs bigger by using a heading font (if that is even possible) and no background color?

CrossEye commented:
New commits:

  - Cleaned out the changed `modification` fields
  - Separated the footer just with white space, removed the color
  - Increased the font-size in the tab buttons
  - Removed the columns in the the tabs.  (Hanging indents was still ugly..)
  - Added a `skip-footer` option, and applied it to all the Welcome tiddlers

I'd love to hear what others think about the icons in the tab buttons.  @yaisog raised an objection, but I think I still prefer it with them, albeit only slightly.  But if we do keep them, then we need to find a better one for the Sources tab.

As to the `skip-footer` mechanism, what I'd really like is for the footer to show on all documentation tiddlers, but not on what me might call marketing ones.  There's no simple query that can distinguish those, is there?

pmario commented:
I think the "Linked here" caption should be "Linked" only. The "here" text is redundant. 

Imo there should be an option for icons only as shown in my screenshot.  I also would like to have the tabs right aligned with icons only. If existing icons do not work, we need new ones. 

For me there should be a separator between the text and the footer even if it is almost transparent. 

Just some thoughts.

pmario commented on editions/tw5.com/tiddlers/$__editions_tw5.com_more-info-view-template.tid:
```diff
@@ -0,0 +1,19 @@
+title: $:/editions/tw5.com/more-info-view-template
+type: text/vnd.tiddlywiki
+
+<% if [<currentTiddler>!is[system]!skip-footer[yes]] %>
+
+<div class="tc-more-info-footer">
+<<tabs 
+  "[tag[$:/tags/MoreInfoFooter]!has[draft.of]]"
+  $:/editions/tw5.com/more-info-tab-related
+  "$:/state/more-info-footer-tab"
+  buttonTemplate:"$:/editions/tw5.com/more-info-tab-button-template" 
+>>
+</div>
+
+<% endif %>
```
This code currently creates a redundant P tag, which should be avoided. 

```
\whitespace trim

<% if [<currentTiddler>!is[system]!skip-footer[yes]] %>
	<div class="tc-more-info-footer">
		<<tabs 
			"[tag[$:/tags/MoreInfoFooter]!has[draft.of]]"
			$:/editions/tw5.com/more-info-tab-related
			"$:/state/more-info-footer-tab"
			buttonTemplate:"$:/editions/tw5.com/more-info-tab-button-template" 
		>>
	</div>
<% endif %>
```

pmario commented on editions/tw5.com/tiddlers/$__editions_tw5.com_more-info-view-template.tid:
```diff
@@ -0,0 +1,19 @@
+
+<% if [<currentTiddler>!is[system]!skip-footer[yes]] %>
+
+<div class="tc-more-info-footer">
```
IMO it should be `class="tc-tiddler-footer"`

CrossEye replied:
I have no problem with that, so long as we want to claim the term for this.

pmario commented on editions/tw5.com/tiddlers/$__editions_tw5.com_more-info-view-template.tid:
```diff
@@ -0,0 +1,19 @@
+list-after: $:/core/ui/ViewTemplate/body
+modified: 20250327020654861
+tags: $:/tags/ViewTemplate
+title: $:/editions/tw5.com/more-info-view-template
```
IMO `title: $:/editions/tw5.com/tiddler-related-template`

CrossEye replied:
I think of that as simply one of the tabs, although I guess it can refer to other things.  The Wikipedia example that @yaisog shared makes me want to go back to "See Also" for that first tab anyway.

pmario commented on editions/tw5.com/tiddlers/system/Sources.tid:
```diff
@@ -1,10 +1,12 @@
+icon: $:/core/images/transcludify
 modified: 20240313115026563
-tags: $:/tags/TiddlerInfo
+tags: $:/tags/TiddlerInfo $:/tags/MoreInfoFooter
```
IMO `$:/tags/MoreInfoFooter` -> `$:/tags/TiddlerFooter` -- IMO the footer should be a new core feature, fist used in the docs

IMO "more" and "info" is where the functionality comes from, but in this case is redundant information and should be removed.

CrossEye replied:
Hmm, I was only thinking about the docs.  A core feature is going to require a *lot* more thinking.

pmario commented on editions/tw5.com/tiddlers/system/tw5.com-styles.tid:
```diff
@@ -164,3 +164,9 @@ type: text/vnd.tiddlywiki
 	background: linear-gradient(#ffd, #ffd) padding-box, repeating-linear-gradient(-45deg, red 0, red 25%, transparent 0, transparent 50%) 0 / .6em .6em;
 }
+
+.tc-more-info-footer {
```
`tc-tiddler-footer`

pmario commented:
I am OK with the existence of a "show footer" configuration, but I am not sure about it's `skip-footer` implementation. 

**Just some brainstorming**

At the moment the tabs are visible, even if there is no info. -> That's OK for me. I am a fan of visible information. But ... many users do not like that. 

So there should be a tab-state, similar to "folded", where is shows the existence of a tab, but it hides the tab body. 

That's why the initial suggestion was icons-only and tabs are "float right".  The style Jeremy and I did discuss (which I can not find atm) was something similar to a style invented by Alberto Molina at http://tw5.scholars.tiddlyspot.com/

It has a "close button", if the info section is visible and it has a "folded state" where it shows the information, that there "is something" but it is very unobtrusive. 

**visible state**

![image](https://github.com/user-attachments/assets/69bfc9c3-14e8-45a7-b8fc-6352401b6690)

**folded state**

![image](https://github.com/user-attachments/assets/cddccedd-cc45-4fcb-a2d9-3f23c478cdb6)

pmario commented:
It seems @kookma has refined the idea for the [Mehregan Edition](https://talk.tiddlywiki.org/t/mehregan-edition-first-public-beta-release/1597) 

If we open the [AdvancedSearch](https://kookma.github.io/TWE-Mehregan/#%24%3A%2FAdvancedSearch) and search for **$:/tags/Thinkup/SmartTab** in the **Shadows** tab, we will get this list: 

[$:/plugins/kookma/thinkup/settings/smart-tabs](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Fsettings%2Fsmart-tabs)
[$:/plugins/kookma/thinkup/stabs/comments](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Fstabs%2Fcomments)
[$:/plugins/kookma/thinkup/stabs/journal-activities](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Fstabs%2Fjournal-activities)
[$:/plugins/kookma/thinkup/stabs/journal-tasks](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Fstabs%2Fjournal-tasks)
[$:/plugins/kookma/thinkup/stabs/keywords](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Fstabs%2Fkeywords)
[$:/plugins/kookma/thinkup/stabs/nodes](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Fstabs%2Fnodes)
[$:/plugins/kookma/thinkup/stabs/todos](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Fstabs%2Ftodos)
[$:/plugins/kookma/thinkup/templates/new-smart-tab](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Ftemplates%2Fnew-smart-tab)
[$:/plugins/kookma/thinkup/viewtemplates/smart-tabs](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Fviewtemplates%2Fsmart-tabs)
[$:/plugins/kookma/thinkup/viewtemplates/smart-tabs-journal](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup%2Fviewtemplates%2Fsmart-tabs-journal)
[$:/tags/Thinkup/SmartTab](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Ftags%2FThinkup%2FSmartTab)

That seems to be the basics of what I have in mind. -- I'll have a closer look

yaisog commented:
Icon-only tabs are pretty useless if the user is on mobile and doesn't have tooltips. Most of these icons mean nothing to me.

There is a setting somewhere whether to show text next to (e.g. ViewTemplate) buttons. Maybe this can be applied here also.

However, that is something that can always be refined later. It is not set in stone. We should concentrate on getting the structure and content right. Otherwise we'll discuss layout and names forever until everyone loses interest to put any more work into it and the whole thing fizzles out like so many other things. Perfect is the enemy of good.

yaisog commented:
I don't believe that this will ever become a core feature. It's simply not universal enough, and it can be easily built by those who want it - meaning no disrespect to @CrossEye and his work on these PRs. =)

The closest you can hope for is for this to become an official plugin, but that also has to play nice (functionally and visually) with e.g. the _Comments_ plugin and whatnot. But this should not be what this PR is about.

For now we should focus on making tiddlywiki.com easier to use and to find information there, not adding all sorts of bells and whistles. Until this PR is merged we cannot even start defining the actual relations between the tiddlers, which will be the real benefit to visitors. As long as this is done only for the tw5-edition, we will still be able to rename the classes and change the icons, colors and words down the road...

CrossEye commented:
@yaisog:
> I don't believe that this will ever become a core feature.

I'm pretty sure I agree.  @pmario surprised me with that, and I tried to adjust my thinking to match.  But I don't think it makes a great deal of sense.

>  It's simply not universal enough, and it can be easily built by those who want it - meaning no disrespect to @CrossEye and his work on these PRs. =)

Absolutely no offense taken.  The implementation is dead-simple.

> Until this PR is merged we cannot even start defining the actual relations between the tiddlers, which will be the real benefit to visitors.

Yes, the rest of the footer was just a nice-to-do-while-I'm-here.  It's the simply declarative assignment of associated tiddlers that I really care about.

> As long as this is done only for the tw5-edition, we will still be able to rename the classes and change the icons, colors and words down the road...

Sure, but I also have no problem renaming them as @pmario suggested; if it make this more future-proof, then I'm all for it.

I did just notice that I'm doing this against the master branch.  I'm pretty sure that this should be against the `tiddlywiki.com` one.  When I make another version, I'll start yet another PR against that branch

Jermolene commented:
Thanks @CrossEye @yaisog @pmario 

I would like to see something along these lines in the core. I have always liked Alberto Molina's "TiddlyWiki for Scholars" at http://tw5.scholars.tiddlyspot.com/ which features bottom tabs that are styled differently from the standard tabs.

<img width="758" alt="image" src="https://github.com/user-attachments/assets/a97efa9f-3928-45cf-9d62-b9cafb99559f" />

We could implement these as another type of tabset (to go along with vertical tabsets). Once that is in place we could choose a small handful of footer tabs that would make sense for the core, and perhaps some further tabs that are specific to  tiddlywiki.com. I am not sure if it would matter if there was overlap with the tabs shown in the info panel. In any event, users would be able to add their own footer tabs.

Conceivably it might be useful to have a mode where the footers of all tiddlers are synchronised together, and always show the same tab.

CrossEye commented:
@Jermolene:

> I would like to see something along these lines in the core.

Can I suggest that we do a first pass in the docs wiki (where we don't need to worry about i18n or creating the perfect abstraction) either trying to get a similar look and feel to *Scholars* or simply holding off on that for the moment?  Then later, when we're happy with the basic functionality, we could extract a version useful in core, tweaking the look and feel, adding whatever i18n necessary, and in general, and trying to make it more universal.

But in the meanwhile, I can continue in my primary purpose, making it easy to add to tiddlers simple links to related info.

As I [explained in `talk`](https://talk.tiddlywiki.org/t/12191/15), my motivating frustration was a more general version of this:

> The docs for `tm-add-field` did not really explain how to use it, and there were no links to other documentation that did explain.

A single link to [ActionSendMessageWidget](https://tiddlywiki.com/#ActionSendMessageWidget), in a `See Also` would make a big difference.  

Once the infrastructure is in place, it should be a relatively straightforward job of going through the docs tiddlers and adding `see-also` (or `related`) fields where there seem to be gaps.

Then we can experiment with new tiddler footers in the docs wiki before trying to add a core feature/plugin.

Does that sound like a reasonable plan?

kookma commented:
> That seems to be the basics of what I have in mind. -- @kookma Do you have a plugin or some code with the "smart tabs" **only**? We do not need all the functionality of the "[thinkup](https://kookma.github.io/TWE-Mehregan/#%24%3A%2Fplugins%2Fkookma%2Fthinkup)" plugin.

Hello @pmario,
SmartTabs are included in Thinkup, and there is no separate plugin. The list of tiddlers you provided is correct. icons were sourced from https://morosanuae.github.io/tw-icons.

yaisog commented:
> Once the infrastructure is in place, it should be a relatively straightforward job of going through the docs tiddlers and adding `see-also` (or `related`) fields where there seem to be gaps.

If we could at least agree on the name of the field, we could start populating it in separate PRs. Then, until we agree on the visual presentation, one could drag-import a set of tiddlers that will show a footer with these related-links into tiddlywiki.com to use them. This could be the current set by @CrossEye, or Jeremy's vision for it, or any other design. Just until there is an official implementation.

Though I still don't understand why we cannot have an initial implementation on the webpage now and then refine that with user feedback, until it eventually makes it into the core. The relationships between the tiddlers themselves and their visual presentation to me are two wholly separate things. I don't want this to become the next breadcrumb PR.

linonetwo commented:
My plugin https://tiddly-gittly.github.io/inverse-link-and-folder/ supports two modes, one is full preview

![图片](https://github.com/user-attachments/assets/57369f72-c5be-4466-b144-3b92bf483a14)

Another is simple mode, maybe you could also add preview mode.

![图片](https://github.com/user-attachments/assets/bb43dd95-32f3-465a-935b-f364a2685fd2)

And don't forget we don't only have backlink, we also have backtransclude.

And there should be i18n when being a core feature.

![图片](https://github.com/user-attachments/assets/66c553f5-2ccd-4904-866f-fa3b2e77ca0f)

yaisog commented:
Hi @CrossEye, if we ever get around to start adding the actual relationships between individual docs tiddlers (which I fear is being sidetracked indefinitely be the discussion on their presentation), I have created a ViewTemplate addition that will show an editor for the `related` field at the bottom of each tiddler. This is based on the `tags` field editor with some minor modifications (mainly to show plain text instead of tag pills). This way we could just add any number of related tiddlers without needing to enter edit mode or to manually edit lists:
![image](https://github.com/user-attachments/assets/b94c907f-4618-40ea-b2c8-c982c7af6b2f)
It's not super-pretty, but functional. One would use it by simply importing this tiddler into the Docs PR Maker, adding relationships where needed, and then deleting the tiddler again before submitting (so it doesn't become part of the PR).
[$__core_ui_ViewTemplate_related.json](https://github.com/user-attachments/files/19613905/__core_ui_ViewTemplate_related.json)

If we at some point agree on a name for the field that suppresses the footer, I'll put in a condition that will not show the editor for these tiddlers.

CrossEye commented:
Sorry, I've been heads-down on another project, and will remain so until we manage to run our political event on April 27.

But I will get back to this afterward.

I like the idea of that template, @yaisog`.  I'm also well in support of separating out the ui from the data collection.

I'll be back!
